cap-primitives/src/std/fs/canonicalize_manually.rs

```diff
@@ -1,22 +1,36 @@
 //! Manual path canonicalization, one component at a time, with manual symlink
 //! resolution, in order to enforce sandboxing.
 
-use crate::fs::{open_manually, OpenOptions};
+use crate::fs::{open_manually, FollowSymlinks, OpenOptions};
 use std::{
     fs, io,
     path::{Path, PathBuf},
 };
 
 /// Implement `canonicalize` by breaking up the path into components and resolving
 /// each component individually, and resolving symbolic links manually.
-pub(crate) fn canonicalize_manually(start: &fs::File, path: &Path) -> io::Result<PathBuf> {
+pub(crate) fn canonicalize_manually_and_follow(
+    start: &fs::File,
+    path: &Path,
+) -> io::Result<PathBuf> {
+    canonicalize_manually(start, path, FollowSymlinks::Yes)
+}
+
+/// The main body of `canonicalize_manually`, which takes an extra `follow`
+/// flag allowing the caller to disable following symlinks in the last
+/// component.
+pub(crate) fn canonicalize_manually(
+    start: &fs::File,
+    path: &Path,
+    follow: FollowSymlinks,
+) -> io::Result<PathBuf> {
     let mut symlink_count = 0;
     let mut canonical_path = PathBuf::new();
 
     if let Err(e) = open_manually(
         start,
         path,
-        OpenOptions::new().read(true),
+        follow.options(OpenOptions::new().read(true)),
```

Comment:
Hmm, I'm not sure about this one. Would it perhaps be more readable if we had a method similar to `OpenOptions::nofollow` but instead of a `bool`, it would take `FollowSymlinks` as an argument?

Reply:
Good idea! I've now done that, and it cleans up several areas of code.